src/vmm/src/builder.rs

```diff
@@ -763,39 +767,568 @@ pub mod tests {
+            cmdline: default_kernel_cmdline(),
+            kernel_file: default_kernel_file(),
+        }
+    }
+
+    fn default_vmm(event_manager: &mut EventManager) -> Vmm {
+        let guest_memory = create_guest_memory(128).unwrap();
+        let kernel_cmdline = default_kernel_cmdline();
+
+        let exit_evt = EventFd::new(libc::EFD_NONBLOCK)
+            .map_err(Error::EventFd)
+            .map_err(StartMicrovmError::Internal)
+            .unwrap();
+
+        let vm = setup_kvm_vm(&guest_memory).unwrap();
+        let mmio_device_manager = default_mmio_device_manager();
+        #[cfg(target_arch = "x86_64")]
+        let pio_device_manager = default_portio_device_manager(event_manager);
+
+        Vmm {
+            events_observer: Some(Box::new(SerialStdin::get())),
+            guest_memory,
+            kernel_cmdline,
+            vcpus_handles: Vec::new(),
+            exit_evt,
+            vm,
+            mmio_device_manager,
+            #[cfg(target_arch = "x86_64")]
+            pio_device_manager,
+            seccomp_level: SECCOMP_LEVEL_ADVANCED,
+        }
+    }
+
+    #[test]
+    fn test_stdin_wrapper() {
+        let wrapper = SerialStdin(io::stdin());
+        assert_eq!(wrapper.as_raw_fd(), io::stdin().as_raw_fd())
+    }
+
+    #[test]
+    #[cfg(target_arch = "x86_64")]
+    fn test_create_vcpus_x86_64() {
+        let vcpu_count = 2;
         let mut event_manager = EventManager::new().expect("Unable to create EventManager");
-        setup_serial_device(
-            &mut event_manager,
-            Box::new(read_handle),
-            Box::new(io::stdout()),
+
+        let guest_memory = create_guest_memory(128).unwrap();
+        let mut vm = setup_kvm_vm(&guest_memory).unwrap();
+        setup_interrupt_controller(&mut vm).unwrap();
+        let vcpu_config = VcpuConfig {
+            vcpu_count,
+            ht_enabled: false,
+            cpu_template: None,
+        };
+        let boot_config = default_boot_config();
+        let entry_addr = load_kernel(&boot_config, &guest_memory).unwrap();
+        let pio_dev_manager = default_portio_device_manager(&mut event_manager);
+        let vcpu_vec = create_vcpus_x86_64(
+            &vm,
+            &vcpu_config,
+            &guest_memory,
+            entry_addr,
+            TimestampUs::default(),
+            &pio_dev_manager.io_bus,
         )
         .unwrap();
+        assert_eq!(vcpu_vec.len(), vcpu_count as usize);
     }
+
+    #[test]
+    #[cfg(target_arch = "aarch64")]
+    fn test_create_vcpus_aarch64() {
+        let guest_memory = create_guest_memory(128).unwrap();
+        let vm = setup_kvm_vm(&guest_memory).unwrap();
+        let vcpu_count = 2;
+
+        let vcpu_config = VcpuConfig {
+            vcpu_count,
+            ht_enabled: false,
+            cpu_template: None,
+        };
+
+        let boot_config = default_boot_config();
+        let entry_addr = load_kernel(&boot_config, &guest_memory).unwrap();
```

Comment:
Same here, you don't need to actually load a kernel for this test, just pass some random `entry_addr`. The vcpu will never be booted in this test anyway.